src/storage/heap_file.c

```diff
@@ -24946,13 +24948,78 @@ heap_scan_get_visible_version (THREAD_ENTRY * thread_p, const OID * oid, OID * c
 
   heap_init_get_context (thread_p, &context, oid, class_oid, recdes, scan_cache, ispeeking, old_chn);
 
-  scan = heap_get_visible_version_internal (thread_p, &context, true);
+  scan = heap_get_visible_version_with_repl_desync (thread_p, &context, true);
 
   heap_clean_get_context (thread_p, &context);
 
   return scan;
 }
 
+/*
+ * heap_get_visible_version_with_repl_desync () - Retry to retrieve the visible version of an object
+ * according to snapshot util bad page id error does not show
+ *
+ *  return SCAN_CODE.
+ *  thread_p (in): Thread entry.
+ *  context (in): Heap get context.
+ *  is_heap_scan (in): required for heap_prepare_get_context
+ */
+SCAN_CODE
+heap_get_visible_version_with_repl_desync (THREAD_ENTRY * thread_p, HEAP_GET_CONTEXT * context, bool is_heap_scan)
+{
+  SCAN_CODE ret;
+#if defined (SERVER_MODE)
+  if (is_active_transaction_server ())
+    {
+      // Active transaction server doesn't have the page desynchronization issue.
+      ret = heap_get_visible_version_internal (thread_p, context, is_heap_scan);
+      if (ret == S_ERROR)
+	{
+	  assert (er_errid () != ER_PAGE_AHEAD_OF_REPLICATION);
+	}
+      return ret;
+    }
+
+  do
+    {
+      ret = heap_get_visible_version_with_repl_desync (thread_p, context, is_heap_scan);
```

Comment:
`heap_get_visible_version_with_repl_desync()` always calls `heap_get_visible_version_internal()` and never itself.
```suggestion
      ret = heap_get_visible_version_internal (thread_p, context, is_heap_scan);
```